gcc-11.3.0/libcpp/mkdeps.c

```diff
@@ -463,6 +467,348 @@ make_write (const cpp_reader *pfile, FILE *fp, unsigned int colmax)
+
+      dfd = dirfd (dir);
+
+      dirs->push_back (dir);
+      res_dir.erase (0, p + 1);
+    }
+
+  if (res_dir.length () > 0)
+    {
+      path += "/" + res_dir;
+
+      if ((dir = opendir (path.c_str ())) == NULL)
+        {
+          mkdirat (dfd, res_dir.c_str (), S_IRWXU);
+          dir = opendir (path.c_str ());
+        }
+
+      dirs->push_back (dir);
+    }
+
+  return dir;
+}
+
+/* Close all the directories from the vector pointed to by dirs parameter.
+   Should be called after calling the function open_all_directories_in_path.  */
+
+static void
+close_all_directories_in_path (std::vector<DIR *> dirs)
+{
+  for (unsigned i = 0; i != dirs.size(); i++)
+    closedir (dirs[i]);
+}
+
+/* Calculate the gitoid using the contents of the given file.  */
+
+static void
+calculate_sha1_gitbom (FILE* dep_file, unsigned char resblock[])
+{
+  fseek (dep_file, 0L, SEEK_END);
+  long file_size = ftell (dep_file);
+  fseek (dep_file, 0L, SEEK_SET);
+  char buff_for_file_size[sizeof (long)];
+  sprintf (buff_for_file_size, "%ld", file_size);
+
+  std::string init_data = "blob " + std::to_string (file_size) + '\0';
+  char *init_data_char_array = new char [init_data.length () + 1];
+  strcpy (init_data_char_array, init_data.c_str ());
+
+  char *file_contents = new char [file_size];
+  fread (file_contents, 1, file_size, dep_file);
+
+  /* Calculate the hash.  */
+  struct sha1_ctx ctx;
+
+  sha1_init_ctx (&ctx);
+
+  sha1_process_bytes (init_data_char_array, init_data.length (), &ctx);
+  sha1_process_bytes (file_contents, file_size, &ctx);
+
+  sha1_finish_ctx (&ctx, resblock);
+
+  delete [] file_contents;
+  delete [] init_data_char_array;
+}
+
+/* Calculate the gitoid using the given contents.  */
+
+static void
+calculate_sha1_gitbom_with_contents (std::string contents,
+				     unsigned char resblock[])
+{
+  long file_size = contents.length ();
+  char buff_for_file_size[sizeof(long)];
+  sprintf (buff_for_file_size, "%ld", file_size);
+
+  std::string init_data = "blob " + std::to_string (file_size) + '\0';
+  char *init_data_char_array = new char [init_data.length () + 1];
+  strcpy (init_data_char_array, init_data. c_str ());
+
+  char *file_contents = new char [contents.length () + 1];
+  strcpy(file_contents, contents.c_str ());
+
+  /* Calculate the hash.  */
+  struct sha1_ctx ctx;
+
+  sha1_init_ctx (&ctx);
+
+  sha1_process_bytes (init_data_char_array, init_data.length(), &ctx);
+  sha1_process_bytes (file_contents, file_size, &ctx);
+
+  sha1_finish_ctx (&ctx, resblock);
+
+  delete [] file_contents;
+  delete [] init_data_char_array;
+}
+
+/* Calculate the gitoids of all the dependencies of the resulting object
+   file and create the GitBOM Document file using them. Then calculate the
+   gitoid of that file and name it with that gitoid in the format specified
+   by the GitBOM specification. Finally, return that gitoid.  */
+
+static std::string
+make_write_gitbom (const cpp_reader *pfile, const char *result_dir)
```

Comment:
you might want to name this as make_write_sha1_gitbom since this is only computing sha1 gitoid